File: src-tauri/src/commands/reports.rs
use sea_orm::DatabaseConnection;
use tauri::State;

use crate::domain::reports::*;
use crate::services;
use crate::utils::AppError;

#[tauri::command]
pub async fn report_sales_daily(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<Vec<DailySalesRow>, AppError> {
    services::reports::daily_sales(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_sales_monthly(
    db: State<'_, DatabaseConnection>,
    year: i32,
) -> Result<Vec<MonthlySalesRow>, AppError> {
    services::reports::monthly_sales(db.inner(), year).await
}

#[tauri::command]
pub async fn report_sales_period(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<PeriodSalesSummary, AppError> {
    services::reports::sales_period(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_sales_receipt(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
    search: String,
) -> Result<ReceiptReport, AppError> {
    services::reports::sales_receipts(db.inner(), start_date, end_date, search).await
}

#[tauri::command]
pub async fn report_payment_methods(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<Vec<PaymentMethodRow>, AppError> {
    services::reports::payment_methods(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_product_sales(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<Vec<ProductSalesRow>, AppError> {
    services::reports::product_sales(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_popular_products(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
    limit: i32,
) -> Result<Vec<PopularProductRow>, AppError> {
    services::reports::popular_products(db.inner(), start_date, end_date, limit).await
}

#[tauri::command]
pub async fn report_returns(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<Vec<ReturnRow>, AppError> {
    services::reports::returns(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_current_stock(
    db: State<'_, DatabaseConnection>,
    search: String,
    filter: String,
) -> Result<CurrentStockReport, AppError> {
    services::reports::current_stock(db.inner(), search, filter).await
}

#[tauri::command]
pub async fn report_losses(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<LossSummary, AppError> {
    services::reports::losses(db.inner(), start_date, end_date).await
}

#[tauri::command]
pub async fn report_cash_flows(
    db: State<'_, DatabaseConnection>,
    start_date: String,
    end_date: String,
) -> Result<CashFlowReportSummary, AppError> {
    services::reports::cash_flows(db.inner(), start_date, end_date).await
}

File: src/lib/api/reports.ts
import type {
  CashFlowReportSummary,
  CurrentStockReport,
  DailySalesRow,
  LossSummary,
  MonthlySalesRow,
  PaymentMethodRow,
  PeriodSalesSummary,
  PopularProductRow,
  ProductSalesRow,
  ReceiptReport,
  ReturnRow,
} from "@/features/reports/types"
import { apiGet } from "./client"

/**
 * The eleven reports.
 *
 * The date window is `start_date`/`end_date` — snake_case, because the Rust
 * `DateRange` these routes share has no `rename_all`. Both halves are local
 * `YYYY-MM-DD` dates; the server turns them into UTC boundaries itself.
 *
 * Every figure here is net of refunds, and a return is subtracted on the day it
 * was taken rather than the day of the sale. That means a payment-method row can
 * arrive with a count of zero and a negative amount, and a product row with a
 * quantity of zero or below, for a period that saw only returns. Those are not
 * broken rows — they are the shape of the answer.
 */

interface DateRange {
  startDate: string
  endDate: string
}

function rangeQuery({ startDate, endDate }: DateRange) {
  return { start_date: startDate, end_date: endDate }
}

export function reportSalesDaily(range: DateRange): Promise<DailySalesRow[]> {
  return apiGet<DailySalesRow[]>("/reports/sales/daily", rangeQuery(range))
}

export function reportSalesMonthly(year: number): Promise<MonthlySalesRow[]> {
  return apiGet<MonthlySalesRow[]>("/reports/sales/monthly", { year })
}

export function reportSalesPeriod(range: DateRange): Promise<PeriodSalesSummary> {
  return apiGet<PeriodSalesSummary>("/reports/sales/period", rangeQuery(range))
}

/**
 * The receipt list now includes receipts whose status is `refunded`. A screen
 * that totals a column has to total `netAmount`, not `totalAmount`, or the
 * returned money is counted as if it were still in the till.
 */
export function reportSalesReceipts(
  range: DateRange,
  search: string
): Promise<ReceiptReport> {
  return apiGet<ReceiptReport>("/reports/sales/receipts", {
    ...rangeQuery(range),
    search,
  })
}

export function reportPaymentMethods(range: DateRange): Promise<PaymentMethodRow[]> {
  return apiGet<PaymentMethodRow[]>("/reports/payment-methods", rangeQuery(range))
}

export function reportProductSales(range: DateRange): Promise<ProductSalesRow[]> {
  return apiGet<ProductSalesRow[]>("/reports/products/sales", rangeQuery(range))
}

export function reportPopularProducts(
  range: DateRange,
  limit: number
): Promise<PopularProductRow[]> {
  return apiGet<PopularProductRow[]>("/reports/products/popular", {
    ...rangeQuery(range),
    limit,
  })
}

export function reportReturns(range: DateRange): Promise<ReturnRow[]> {
  return apiGet<ReturnRow[]>("/reports/returns", rangeQuery(range))
}

export function reportCurrentStock(
  search: string,
  filter: string
): Promise<CurrentStockReport> {
  return apiGet<CurrentStockReport>("/reports/stock/current", { search, filter })
}

export function reportLosses(range: DateRange): Promise<LossSummary> {
  return apiGet<LossSummary>("/reports/losses", rangeQuery(range))
}

export function reportCashFlows(range: DateRange): Promise<CashFlowReportSummary> {
  return apiGet<CashFlowReportSummary>("/reports/cash-flows", rangeQuery(range))
}
